### DISCUSSION TITLE
PWA vs stores

### DISCUSSION THREAD

**Discussion by adrienlacombe** [keep-starknet-strange/joyboy#36] [Ideas]
Pros of PWA
- cross platform compatibility
- easier to dev when it comes to Google/Apple specifics
- easier to push/update as no relationship needed with stores
- lighter as heavy lifting is done by browser

Cons of PWA
- Some device features might not be accessed (e.g. Bluetooth?)
- no listing in app stores which is the usual channel users are used to search for apps
- UX is the one of a browser rather than a native app so probably not as slick

Please add if I forgot smth, ty

**Comment by adrienlacombe** [comment]
Expanding on Github CI, when you want to push an app to the stores, you will have to setup builds in GH to build the artefacts. This is not needed in PWA. It might not be a nightmare to maintained on the dev side, but the relationship with the stores might get difficult as we are dealing with crypto.

**Comment by adrienlacombe** [comment]
Maybe biometrics can't be accessed by PWA, investigating

**Reply by AbdelStark** [reply]
Yes, we need to check if we can have access to biometrics / Face ID / keychain / keystore etc

**Reply by w5pand** [reply]
The answer is yes, you can take a look at JoyID wallet, passkey-based encryption wallet, PWA format.
https://joy.id/
https://app.joy.id/

**Reply by AbdelStark** [reply]
Interesting.
Also funny that's the example you mention is also a `Joy` something project.

**Comment by apoorvsadana** [comment]
I have worked on PWAs as apps so I would like to add

1. You can technically list them on stores by making a basic app with the sole purpose of launching a WebView (which contains your website). Android calls this TWA (Trusted Web Activity)
2. In my experience of building on Android, the experience was pretty smooth and the users didn't really know they were actually using a website. In face, many popular apps today also use webviews inside their apps for specific pages, users hardly find out. We didn't use a lot of fancy native features though as they weren't needed.
3. PWAs also save the effort of having a separate team for the web app as long as you build it in a responsive fashion
4. The storage of PWAs was a little inconsistent from what I can recall. If users cleared their browser storage, there app storage (which uses the PWA) also gets reset so it can lead to logouts. It could have been an implementation bug but in general there are small inconsistencies you will find in PWA experiences which you might need to tackle (and it would be less documented compared to normal apps)

**Comment by 0xDeka** [comment]
You should take a look it can be useful
https://stackoverflow.com/questions/63112565/can-mobile-website-have-biometric-authentication

**Comment by thomasmorice** [comment]
I've been implementing both Native and PWA, and I personally think the hassle of taking Native route is worth it, here's why. 

The good about PWA: 
- Quick dev, quick deliveries
- More availabilities for "frontend web devs" than mobile devs
- Developer experience is really nice

The bad about PWA: 
- Missing many features, or requires a lot of headaches to make things work (push notifications for instance) 
- The overall feeling user gets having a webview instead of native components
- Deploying an app is also a headache, store are very severe when it comes to validation

For native dev, and if you are coming from the frontend web development environment, [expo](https://expo.dev/) with react native has been a game changer. The developer experience is better than any competitor, same codebase for android and ios, with expo v50 you can even create api routes (similar to Next.js for instance) and deploy that somewhere so your app can leverage that. 

Expo is the best option right now I would say :)

Cheers 😃

**Comment by ayushtom** [comment]
Just a note - 
with codepush in react native, you can push majority of the changes related to react native over the air without the users ever needing to update the app and we never having to submit the app to playstore to update the app. 
in an expo project like ours - it uses https://docs.expo.dev/eas-update/introduction/ for the same thing
Only native level changes for android and ios specifically which are normally very very few in a react native app, require you to manually update the app by pushing the build through appstores



https://github.com/microsoft/react-native-code-push

**Comment by 0xMikado** [comment]
tl;dr of my personal experience: native web3 app has bad devX and fear of censorship + tax  

Native apps cons: 
- Harder to build, most of the web3 libraries aren't compatible (viem, ethers, privy, rainbow connect,..) But it's getting better, privy I believe is now supported, and others are rolling out. It's hard to have only one codebase to rule them all (android, ios and desktop) 
- You need to pay 30% tax on every transaction.  
- You can be banned whenever the big guys decide they don't want you on the app anymore, pwa can't be banned, it's just the browser  

Other pros of PWAs: 
- Android is pushing to make them 1st class citizens, equal to native apps. IOS doesn't want them but they kinda need to keep them. They tried to remove pwa a few weeks ago in Europe and quickly rolled back their decision (idk why) 
-You can have access to biometric auth and passkey
